pkg/gather/clusterconfig/operators.go

```diff
@@ -92,28 +96,120 @@ func GatherClusterOperators(g *Gatherer, c chan<- gatherResult) {
-	c <- gatherResult{records, errors}
+	records, errs := gatherClusterOperators(g.ctx, gatherConfigClient, gatherKubeClient.CoreV1(), discoveryClient, dynamicClient)
+	c <- gatherResult{records, []error{errs}}
 }
 
-func gatherClusterOperators(ctx context.Context, configClient configv1client.ConfigV1Interface, coreClient corev1client.CoreV1Interface, discoveryClient discovery.DiscoveryInterface, dynamicClient dynamic.Interface) ([]record.Record, []error) {
+// gatherClusterOperators collects cluster operators
+func gatherClusterOperators(ctx context.Context, configClient configv1client.ConfigV1Interface, coreClient corev1client.CoreV1Interface, discoveryClient discovery.DiscoveryInterface, dynamicClient dynamic.Interface) ([]record.Record, error) {
 	config, err := configClient.ClusterOperators().List(ctx, metav1.ListOptions{})
 	if errors.IsNotFound(err) {
 		return nil, nil
 	}
 	if err != nil {
-		return nil, []error{err}
+		return nil, err
 	}
+
+	// collect the cluster operators reports
+	records := clusterOperatorsRecords(ctx, config.Items, dynamicClient, discoveryClient)
+
+	// gather pods from unhealthy operators
+	pods, recs, totalContainers := gatherUnhealthyClusterOperator(ctx, config.Items, coreClient)
+	records = append(records, recs...)
+	// Exit early if no pods found
+	if len(pods) == 0 {
+		return records, nil
+	}
+
+	// gather pods containers logs
+	clogs, err := gatherPodContainersLogs(ctx, coreClient, pods, totalContainers)
+	if err != nil {
+		return records, nil
+	}
+	if len(clogs) > 0 {
+		records = append(records, clogs...)
+	}
+
+	return records, nil
+}
+
+// gatherPodContainersLogs collect the pod current and previous containers logs
+func gatherPodContainersLogs(ctx context.Context, client corev1client.CoreV1Interface, pods []*corev1.Pod, total int) ([]record.Record, error) {
+	var records []record.Record
+
+	if total == 0 {
+		return nil, fmt.Errorf("Total can't be zero.")
+	}
+
+	// Fetch a list of containers in pods and calculate a log size quota
+	// Total log size must not exceed maxLogsSize multiplied by logCompressionRatio
+	klog.V(2).Infof("Found %d pods", len(pods))
+	bufferSize := int64(recorder.MaxLogSize * logCompressionRatio / total / 2)
+	klog.V(2).Infof("Maximum buffer size: %v bytes", bufferSize)
+	buf := bytes.NewBuffer(make([]byte, 0, bufferSize))
+
+	// Fetch previous and current container logs
+	for _, isPrevious := range []bool{true, false} {
+		for _, pod := range pods {
+			clog := collectPodContainerLogs(ctx, client, pod, isPrevious, buf, bufferSize)
+			if len(clog) > 0 {
+				records = append(records, clog...)
+			}
+		}
+	}
+
+	return records, nil
+}
+
+// collectPodContainerLogs generates the previous and current log reports for pod containers using the k8s API response
+func collectPodContainerLogs(ctx context.Context, client corev1client.CoreV1Interface, pod *corev1.Pod, isPrevious bool, buf *bytes.Buffer, bufferSize int64) []record.Record {
```

Comment:
It seems to me little bit confusing having `collectPodContainerLogs` and `gatherPodContainersLogs` functions. The names are almost the same.

Reply:
@tremes any suggestion?

Reply:
Not sure. Either merge these two functions or distinguish the naming.

Reply:
If we merge those functions we gonna increase its code complexity. They are already splitter to reduce it. Let me find a better name so...